File: libspi/streamablecontent.c
/*
 * AT-SPI - Assistive Technology Service Provider Interface
 * (Gnome Accessibility Project; http://developer.gnome.org/projects/gap)
 *
 * Copyright 2001, 2002 Sun Microsystems Inc.,
 * Copyright 2001, 2002 Ximian, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/* streamablecontent.c : implements the StreamableContent interface */

#include <config.h>
#include <stdio.h>
#include <libspi/accessible.h>
#include <libspi/streamablecontent.h>

/* Our parent Gtk object type */
#define PARENT_TYPE SPI_TYPE_BASE

/* A pointer to our parent object class */
static GObjectClass *spi_streamable_parent_class;

static AtkStreamableContent *
get_streamable_from_servant (PortableServer_Servant servant)
{
  SpiBase *object = SPI_BASE (bonobo_object_from_servant (servant));
  g_return_val_if_fail (object != NULL, NULL);
  g_return_val_if_fail (ATK_IS_STREAMABLE_CONTENT(object->gobj), NULL);
  return ATK_STREAMABLE_CONTENT (object->gobj);
}

/*
 * CORBA Accessibility::StreamableContent::getContentTypes method implementation
 */
static Accessibility_StringSeq*
impl_accessibility_streamable_get_content_types (PortableServer_Servant servant,
						 CORBA_Environment     *ev)
{
  Accessibility_StringSeq *typelist = Accessibility_StringSeq__alloc ();
  AtkStreamableContent *streamable = get_streamable_from_servant (servant);
  int n_types, i;

  typelist->_length = 0;
  g_return_val_if_fail (streamable != NULL, typelist);

  n_types = atk_streamable_content_get_n_mime_types (streamable);
  typelist->_length = n_types;
  typelist->_buffer = Accessibility_StringSeq_allocbuf (n_types);
  for (i = 0; i < n_types; ++i) {
    const gchar *mimetype = atk_streamable_content_get_mime_type (streamable, i);
    typelist->_buffer[i] = CORBA_string_dup (mimetype ? mimetype : "");
  }

  return typelist;
}

/*
 * CORBA Accessibility::StreamableContent::getContent method implementation
 */
static Bonobo_Stream
impl_accessibility_streamable_get_content (PortableServer_Servant servant,
					   const CORBA_char * content_type,
					   CORBA_Environment     *ev)
{
  Bonobo_Stream stream;
  AtkStreamableContent *streamable = get_streamable_from_servant (servant);
  GIOChannel *gio;

  g_return_val_if_fail (streamable != NULL, NULL);

  gio = atk_streamable_content_get_stream (streamable, content_type);

  stream = CORBA_OBJECT_NIL; /* FIXME! */

  return stream;
}

static void
spi_streamable_class_init (SpiStreamableClass *klass)
{
        POA_Accessibility_StreamableContent__epv *epv = &klass->epv;
        spi_streamable_parent_class = g_type_class_peek_parent (klass);

        epv->getContentTypes = impl_accessibility_streamable_get_content_types;
        epv->getContent = impl_accessibility_streamable_get_content;
}

static void
spi_streamable_init (SpiStreamable *streamable)
{
}

BONOBO_TYPE_FUNC_FULL (SpiStreamable,
		       Accessibility_StreamableContent,
		       PARENT_TYPE,
		       spi_streamable)

SpiStreamable *
spi_streamable_interface_new (AtkObject *o)
{
    SpiStreamable *retval = g_object_new (SPI_STREAMABLE_TYPE, NULL);

    spi_base_construct (SPI_BASE (retval), G_OBJECT(o));

    return retval;
}

File: pyatspi/streamablecontent.py
#Copyright (C) 2008 Codethink Ltd

#This library is free software; you can redistribute it and/or
#modify it under the terms of the GNU Lesser General Public
#License version 2 as published by the Free Software Foundation.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU Lesser General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import interfaces
from base import BaseProxy, Enum
from factory import add_accessible_class

__all__ = [
	   "ContentStream",
	   "StreamableContent",
	  ]

#------------------------------------------------------------------------------

class ContentStream(BaseProxy):
    """
    An interface by which the requested data from a StreamableContent
    object may be read by the client.
    """
    
    def close(self, *args, **kwargs):
        """
        close the stream and release associated resources. A client should
        not perform further operations on a StreamableContent::Stream
        object after closing it.
        """
        func = self.get_dbus_method("close")
        return func(*args, **kwargs)
    
    def read(self, *args, **kwargs):
        """
        Request/read a specified amount of data from a Stream. 
        @return the number of bytes actually read into the client buffer.
        """
        func = self.get_dbus_method("read")
        return func(*args, **kwargs)
    
    def seek(self, *args, **kwargs):
        """
        Seek to a specified position in the Stream. 
        @param : offset
        an offset specifying the requested position in the stream, relative
        to the SeekType specified in whence. 
        @param : whence
        a SeekType specifying the reference point from which the seek
        offset is calculated. Some forms of seek are not supported by
        certain implementations of Stream, in which case a NotSupported
        exception will be raised. 
        @return the actual resulting offset, if no exception was raised.
        """
        func = self.get_dbus_method("seek")
        return func(*args, **kwargs)
    
    class IOError(Exception):
        pass
    
    class NoPermission(Exception):
        pass
    
    class NotSupported(Exception):
        pass

    class SeekType(Enum):
        """
        Specifies the meaning of a seek 'offset'. Not all SeekTypes are
        supported by all StreamableContent data sources, for instance
        some streams may not support seeking from the beginning or other
        types of 'backwards' seeks.
        """
        _enum_lookup = {
            0:'SEEK_SET',
            1:'SEEK_CURRENT',
            2:'SEEK_END',
        }
    
    SEEK_CURRENT = SeekType(1)
    
    SEEK_END = SeekType(2)
    
    SEEK_SET = SeekType(0)

#------------------------------------------------------------------------------

class StreamableContent(BaseProxy):
    """
    An interface whereby an object allows its backing content to
    be streamed to clients. Negotiation of content type is allowed.
    Clients may examine the backing data and transform, convert,
    or parse the content in order to present it in an alternate form
    to end-users.
    """
    
    def getContent(self, *args, **kwargs):
        """
        DEPRECATED, use getStream instead. getContent: Retrieve this
        object's content, in a format appropriate to a requested mimetype.
                long Bonobo::Stream:seek (in long offset, in SeekType
        whence)
                        raises (NoPermission, IOError)
                void Bonobo::Stream:read (in long count, out iobuf buffer)
                        raises (NoPermission, IOError)
               
        @return a Bonobo::Stream whose mimetype matches contentType,
        if available, or NIL.
        """
        func = self.get_dbus_method("getContent")
        return func(*args, **kwargs)
    
    def getContentTypes(self, *args, **kwargs):
        """
        getContentTypes: 
        @return the list of available mimetypes for this object's content.
        """
        func = self.get_dbus_method("getContentTypes")
        return func(*args, **kwargs)
    
    def getStream(self, *args, **kwargs):
        """
        Retrieve this object's content, in a format appropriate to a
        requested mimetype, as a ContentStream instance.
        @param : contentType
        a string specifying the desired mimetype for the content stream.
        @return a Stream whose mimetype matches contentType, if available,
        or NIL.
        """
        func = self.get_dbus_method("getStream")
        return func(*args, **kwargs)
    
    def getURI(self, *args, **kwargs):
        """
        Get a URI pointing to the content of the specified type, if such
        a URI can be obtained. Not all streamable content providers have
        URI representations.
        @param : contentType
        a string specifying the desired mimetype for the content stream.
        If NULL, then a URI for the default content type will be returned,
        if available.
        @return a string which constitutes a URI for a stream of the
        specified content type, or NULL if no such URI can be obtained.
        """
        func = self.get_dbus_method("getURI")
        return func(*args, **kwargs)
    
# ATTENTION - Register the Application class with the accessible factory.
add_accessible_class(interfaces.ATSPI_STREAMABLE_CONTENT, StreamableContent)

#END----------------------------------------------------------------------------
